Openemr fhir observations for vitals and smoking status

adunsulag opened the pull request:
This implements ONC requirements for all the vital-sign observations as well as the smoking-status. 

Implemented more of the missing FHIR resources per #4205.

Added the pediatric bmi, head-length, & head circumference percentile fields for both the vitals_form and for the FHIR resources.  vital_forms pediatric fields only show if user's age is 20 or younger.  Fixed all of the implementations for FHIR vitals to pass ONC certification.

Added a uuid to the form_vitals.  Got the form vitals to populate in the uuid_mapping table.  Also added a resourcePath that lets us map a FHIR resource observation to a particular search path.

![image](https://user-images.githubusercontent.com/228117/121437179-de0e2380-c94f-11eb-8c86-19afabd658ed.png)
![image](https://user-images.githubusercontent.com/228117/121437225-efefc680-c94f-11eb-8909-40498eac2ce2.png)

Everything is testing well.  There's one minor problem in the inferno testing tool with Blood Pressure, but I'm investigating the testing tool as that appears to be the problem rather than the implementation.  More testing will determine if that's correct or not.

bradymiller commented on interface/forms/vitals/FormVitals.class.php:
```diff
@@ -326,8 +334,84 @@ public function set_oxygen_flow_rate($o)
+    }
     public function persist()
     {
+        if (empty($uuid)) {
```
should this be checking `$this->uuid`

adunsulag replied:
yes, fixing.

bradymiller commented on interface/forms/vitals/FormVitals.class.php:
```diff
@@ -326,8 +334,84 @@ public function set_oxygen_flow_rate($o)
+            $this->uuid = (new UuidRegistry(['table_name' => self::TABLE_NAME]))->createUuid();
+        }
         parent::persist();
+        $fhirVitalsService = new \OpenEMR\Services\FHIR\FhirVitalsService();
```
Should this be in the `use` block?

adunsulag replied:
This should be fixed now.

bradymiller commented on interface/forms/vitals/FormVitals.class.php:
```diff
@@ -326,8 +334,84 @@ public function set_oxygen_flow_rate($o)
         parent::persist();
+        $fhirVitalsService = new \OpenEMR\Services\FHIR\FhirVitalsService();
+        // TODO: @adunsulag we should really make this so it populates it for just the one uuid we make..
+
```
placemark for ongoing code review

bradymiller replied:
still here (sleep time, but should have time to review this tomorrow night :) )

bradymiller commented on src/Services/FHIR/FhirVitalsService.php:
```diff
@@ -0,0 +1,698 @@
+            // TODO: @adunsulag make this a single function call so we can be more effecient
+//            $resourcePathList[] = "category=vital-signs&code=" . $mapping['code'];
+            $resourcePath = $this->getResourcePathForCode($mapping['code']);
+            UuidMapping::createMissingResourceUuids('Observation', 'form_vitals', $resourcePath);
```
![2DV (1)](https://user-images.githubusercontent.com/278968/121653587-b4d5bc00-ca51-11eb-8b90-c57d056eba59.gif)

bradymiller commented on sql/6_0_0-to-6_1_0_upgrade.sql:
```diff
@@ -679,3 +679,27 @@ INSERT INTO `list_options` (`list_id`, `option_id`, `title`, `seq`, `is_default`
+#EndIf
+
+#IfMissingColumn form_vitals ped_head_circ
+ALTER TABLE `form_vitals` ADD `ped_head_circ` FLOAT(4,1) NULL DEFAULT '0.00'
```
just noting difference where NULL us in these 3 percentile entries, but not in the corresponding database.sql entried

adunsulag replied:
This should be fixed.

bradymiller commented on sql/6_0_0-to-6_1_0_upgrade.sql:
```diff
@@ -679,3 +679,27 @@ INSERT INTO `list_options` (`list_id`, `option_id`, `title`, `seq`, `is_default`
+#IfMissingColumn form_vitals uuid
+ALTER TABLE `form_vitals` ADD `uuid` binary(16) DEFAULT NULL AFTER `id`;
+#EndIf
+
```
```
#IfNotIndex form_vitals uuid
CREATE UNIQUE INDEX `uuid` ON `form_vitals` (`uuid`);
#EndIf

#IfUuidNeedUpdate form_vitals
#EndIf
```

adunsulag replied:
Should be fixed now.

bradymiller replied:
this still needs to be addressed

bradymiller commented on sql/6_0_0-to-6_1_0_upgrade.sql:
```diff
@@ -679,3 +679,27 @@ INSERT INTO `list_options` (`list_id`, `option_id`, `title`, `seq`, `is_default`
+
+#IfMissingColumn history_data uuid
+ALTER TABLE `history_data` ADD `uuid` binary(16) DEFAULT NULL AFTER `id`;
+#EndIf
```
```
#IfNotIndex history_data uuid
CREATE UNIQUE INDEX `uuid` ON `history_data` (`uuid`);
#EndIf

#IfUuidNeedUpdate history_data
#EndIf
```

adunsulag replied:
Should be fixed now.

bradymiller commented on sql/database.sql:
```diff
@@ -2187,6 +2188,9 @@ CREATE TABLE `form_vitals` (
+  `ped_weight_height` float(4,1) default '0.0',
+  `ped_bmi` float(4,1) default '0.0',
+  `ped_head_circ` float(4,1) default '0.0',
   PRIMARY KEY  (`id`),
```
```
UNIQUE KEY `uuid` (`uuid`),
```

adunsulag replied:
Should now be fixed.

bradymiller commented on sql/database.sql:
```diff
@@ -8518,6 +8523,7 @@ CREATE TABLE `uuid_mapping` (
+  `resource_path` VARCHAR(255) DEFAULT NULL,
   `table` varchar(255) NOT NULL DEFAULT '',
   `target_uuid` binary(16) NOT NULL DEFAULT '',
   `created` timestamp NULL,
```
need following key in this table:
```
UNIQUE KEY `uuid` (`uuid`),
```

adunsulag replied:
I believe I fixed this now.

adunsulag commented:
In addition to the observation labs, I've included the following:

Fixed consultation_note LOINC code being incorrect
Implemented DocumentReference implementation for Clinical Notes and Patient Documents.  (#4205)
Patient Document access still has a number of issues such as LOINC categorization and UUID implementation.
Added search capability to base DocumentService.  
Put in a stubbed implementation of DiagnosticReport (#4307)
Added the base fhir url to the FhirServiceBase so we can make FQDN urls for our Document Reference class.



In terms of what's outstanding for the FHIR, we have the following:
**Single Patient API**

Missing Resources (woohoo down to just a couple)
	Diagnostic Report resource
	Procedures
	MedicationRequest
	Observation - Laboratory Results
	Provenance resource -- standalone resource endpoint


Fix Implementation
	Encounter -- missing elements
	Organization -- access right checks
	Practitioners -- access right checks
	CareTeam - team status flag
	DocumentReference - uuid for documents instead of device_uuid
	Observation - blood pressure figure out FHIR bug

API cleanup
	-- fix some general todos and other cleanup

**Multi-Patient API**
	For everything but patient implement the system/group/patient export for each of the resources
	These should go very fast (see FhirPatientService->export() as an example).

Visual Check
	Need to implement Patient previous name in FHIR
	Implement Token Invalidation

Additional Improvements
	Remember EHR-Launch practitioner login so login is not required again

**API Documentation**

Look at all the green checkmarks haha:
![image](https://user-images.githubusercontent.com/228117/122454555-87b96a00-cf79-11eb-8830-c9fea19b6c9b.png)
![image](https://user-images.githubusercontent.com/228117/122454614-9869e000-cf79-11eb-99c5-64a016025a02.png)
![image](https://user-images.githubusercontent.com/228117/122454675-a4ee3880-cf79-11eb-9ec3-b91f42971be0.png)

adunsulag commented:
@bradymiller I noticed that I had commited everything with the wrong email address, so I rebased it all to fix the email address problem.  I noticed that the commit links on some of your review comments are missing.  Hopefully that doesn't mess you up.  Just wanted to give you a heads up though if you scratch your head wondering what happened.

bradymiller commented on src/Services/ClinicalNotesService.php:
```diff
@@ -0,0 +1,299 @@
+use OpenEMR\Services\Search\ISearchField;
+use OpenEMR\Services\Search\SearchFieldException;
+use OpenEMR\Validators\ProcessingResult;
+use Ramsey\Uuid\Rfc4122\UuidV4;
```
above line can be removed

bradymiller commented on src/Services/FHIR/DocumentReference/FhirPatientDocumentReferenceService.php:
```diff
@@ -0,0 +1,211 @@
+
+        $id = new FHIRId();
+        // TODO: @adunsulag after playing around with this we definitely need a document uuid, drive_uuid doesn't work
+        // as it only applies to anything not couchdb...
```
agreed, as i recall the drive_uuid is just meant to be the name of the actual filename stored on the drive or couchdb. would not use it as the official uuid identifier.

adunsulag replied:
should be resolved now.

bradymiller commented on src/Services/FHIR/FhirSocialHistoryService.php:
```diff
@@ -0,0 +1,379 @@
+
+    private function parseSocialHistoryIntoObservationRecords(ProcessingResult $processingResult, $record, $observationCodesToReturn)
+    {
+        $uuidMappings = $this->getUuidMappings(UuidRegistry::uuidToBytes($record['uuid']));
```
![2DV](https://user-images.githubusercontent.com/278968/122898574-d8292280-d2ff-11eb-844b-a5bdb1f16ca8.gif)

bradymiller commented on interface/forms/clinical_notes/table.sql:
```diff
@@ -4,6 +4,8 @@
+  `form_id` bigint(20) NOT NULL,
   `date` DATE DEFAULT NULL,
   `pid` bigint(20) DEFAULT NULL,
   `encounter` varchar(255) DEFAULT NULL,
```
```
UNIQUE KEY `uuid` (`uuid`)
```

adunsulag replied:
This should be fixed now.

bradymiller commented on interface/forms/vitals/table.sql:
```diff
@@ -20,5 +21,9 @@ CREATE TABLE IF NOT EXISTS `form_vitals` (
+`ped_weight_height` FLOAT(4,1)      DEFAULT '0.00' NULL,
+`ped_bmi`           FLOAT(4,1)      DEFAULT '0.00' NULL,
+`ped_head_circ`     FLOAT(4,1)      DEFAULT '0.00' NULL,
 PRIMARY KEY (id)
```
```
UNIQUE KEY `uuid` (`uuid`)
```

adunsulag replied:
This should be fixed now.

bradymiller commented on sql/6_0_0-to-6_1_0_upgrade.sql:
```diff
@@ -592,7 +592,7 @@ INSERT INTO `list_options` (`list_id`, `option_id`, `title`, `seq`, `is_default`
 INSERT INTO `list_options` (`list_id`, `option_id`, `title`, `seq`, `is_default`, `option_value`, `mapping`, `notes`, `codes`, `toggle_setting_1`, `toggle_setting_2`, `activity`, `subtype`, `edit_options`) VALUES ('Clinical_Note_Type','discharge_summary','Discharge Summary Note',50,0,0,'','LOINC:18842-5','',0,0,1,'',1);
 INSERT INTO `list_options` (`list_id`, `option_id`, `title`, `seq`, `is_default`, `option_value`, `mapping`, `notes`, `codes`, `toggle_setting_1`, `toggle_setting_2`, `activity`, `subtype`, `edit_options`) VALUES ('Clinical_Note_Type','procedure_note','Procedure Note',60,0,0,'','LOINC:28570-0','',0,0,1,'',1);
-INSERT INTO `list_options` (`list_id`, `option_id`, `title`, `seq`, `is_default`, `option_value`, `mapping`, `notes`, `codes`, `toggle_setting_1`, `toggle_setting_2`, `activity`, `subtype`, `edit_options`) VALUES ('Clinical_Note_Type','consultation_note','Consultation Note',70,0,0,'','LOINC:81222-2','',0,0,1,'',1);
+INSERT INTO `list_options` (`list_id`, `option_id`, `title`, `seq`, `is_default`, `option_value`, `mapping`, `notes`, `codes`, `toggle_setting_1`, `toggle_setting_2`, `activity`, `subtype`, `edit_options`) VALUES ('Clinical_Note_Type','consultation_note','Consultation Note',70,0,0,'','LOINC:11488-4','',0,0,1,'',1);
```
should instead change this using a #IfRow3D call (at end of this script)

adunsulag replied:
I fixed this and added it at the end of the script.

bradymiller commented on sql/6_0_0-to-6_1_0_upgrade.sql:
```diff
@@ -687,3 +687,63 @@ ALTER TABLE `form_clinical_notes` ADD `note_related_to` TEXT COMMENT 'Reference
+CREATE INDEX `uuid` ON `history_data` (`uuid`);
+#EndIf
+
+#IfUuidNeedUpdate history_data
```
@adunsulag , Just trying to figure out best way to go here since per above these are not going to be a uuid per row. If this is instead a unique uuid per patient, then would use the vertical table update mechanism (#IfUuidNeedUpdateVertical) with the pid as the grouping index.

adunsulag replied:
The problem is there is nothing unique in the history_data table so the vertical table update doesn't work as there is no true composite key where the combination of columns is unique.  What we may need to do is rearchitect the table structure and have a sub table such as history_data_archive or something like that if we truly need the historical audit of the data.

bradymiller commented on sql/database.sql:
```diff
@@ -1748,7 +1752,8 @@ CREATE TABLE `form_clinical_notes` (
     `clinical_notes_type` varchar(100) DEFAULT NULL,
-    `note_related_to` text
+    `note_related_to` text,
+    PRIMARY KEY (`id`)
```
```
UNIQUE KEY `uuid` (`uuid`)
```

bradymiller commented on sql/database.sql:
```diff
@@ -2188,6 +2194,9 @@ CREATE TABLE `form_vitals` (
+  `ped_bmi` float(4,1) default '0.0',
+  `ped_head_circ` float(4,1) default '0.0',
   PRIMARY KEY  (`id`),
   KEY `pid` (`pid`)
```
```
UNIQUE KEY `uuid` (`uuid`)
```

bradymiller commented on src/Services/ObservationLabService.php:
```diff
@@ -38,6 +44,143 @@ public function __construct()
+
+    private function getSampleLaboratoryResults()
+    {
+        $factory = new \Ramsey\Uuid\UuidFactory();
```
would include this in use block above just to make easier to track class use

adunsulag replied:
this should be fixed now.

bradymiller commented:
@adunsulag , review complete :)
Regarding history_data, will leave it up to you on whether to just treat uuid as vertical mapping to pid or if wish to remove all the old data and make it a standard table with unique uuid (and pid).

adunsulag commented:
@bradymiller so I did a bunch of searching and didn't see where the history_data was used anywhere and got about halfway through the archive table implementation and then I found where its used.  The historical table with date ranges is used in the clinical decision rules code.  So I'm going to back out my changes and go with the vertical mapping on the pid.  I could probably make the archive table work but it is becoming a bunch of work so maybe we can come back to this another time.

sjpadgett commented:
A boo boo in code_types.inc L-78 I wonder why my PR didn't catch this yesterday?
```
$$GLOBALS['default_search_code_type'] = key($code_types);
```

sjpadgett commented:
nope I was just looking at the failed test.

stephenwaite commented:
funny my [PR](https://github.com/openemr/openemr/commit/f0037dbaf1cd2e4642779201fc107ee6382dda64) went through ok too

stephenwaite commented:
![signal-2021-01-12-153128](https://user-images.githubusercontent.com/5618372/123809210-6cc2f000-d8bf-11eb-82bb-779f7defd985.gif)

adunsulag commented:
@sjpadgett At least one of the tests is failing because I'm using date as a vertical column option which fails the uuid population.  Strangely enough the code executing on github is not what is happening on my local box.  I'm going to try rebasing and see if I can resolve that issue, but I'll have to wait on @stephenwaite to get his fix in.  Oh well, I'll just keep on working on DiagnosticReport... oh goodie.

sjpadgett commented:
@adunsulag That still needs to be fixed and rebasing prob won't do it. Syntax is wrong and I think the key(var) may need to be cast or validated first.

bradymiller commented on interface/forms/clinical_notes/table.sql:
```diff
@@ -4,6 +4,8 @@
+  `form_id` bigint(20) NOT NULL,
   `date` DATE DEFAULT NULL,
   `pid` bigint(20) DEFAULT NULL,
   `encounter` varchar(255) DEFAULT NULL,
```
```
UNIQUE KEY `uuid` (`uuid`)
```

adunsulag replied:
done

bradymiller commented on interface/forms/vitals/table.sql:
```diff
@@ -20,5 +21,10 @@ CREATE TABLE IF NOT EXISTS `form_vitals` (
+`oxygen_flow_rate`  FLOAT(5,2)      DEFAULT '0.00' NULL,
+`ped_weight_height` FLOAT(4,1)      DEFAULT '0.00' NULL,
+`ped_bmi`           FLOAT(4,1)      DEFAULT '0.00' NULL,
+`ped_head_circ`     FLOAT(4,1)      DEFAULT '0.00' NULL,
```
remove NULL in the 4 lines above

adunsulag replied:
done

bradymiller commented on sql/6_0_0-to-6_1_0_upgrade.sql:
```diff
@@ -687,3 +687,87 @@ ALTER TABLE `form_clinical_notes` ADD `note_related_to` TEXT COMMENT 'Reference
+#IfNotIndex history_data uuid
+-- History table uses a historical record data pattern where the same record is duplicated multiple times to retain
+-- history, this makes it so the record has the same uuid.  We only index the record instead of keeping it UNIQUE
+CREATE UNIQUE INDEX `uuid` ON `history_data` (`uuid`);
```
if we are back to a unique index, haven't we gone full circle to treat it like the other tables in terms of uuid population?
(ie. just using `#IfUuidNeedUpdate history_data` in addition to not using vertical table calls when populating uuids for it elsewhere in codebase)

adunsulag replied:
It does look like we've gone back full circle due to the clinical decision rules treating each history_data record as its own observation.  I changed all of the history_data to have its own uuid and removed the vertical_table stuff.

bradymiller commented on sql/6_0_0-to-6_1_0_upgrade.sql:
```diff
@@ -687,3 +687,87 @@ ALTER TABLE `form_clinical_notes` ADD `note_related_to` TEXT COMMENT 'Reference
+#EndIf
+
+#IfMissingColumn form_clinical_notes clinical_notes_category
+ALTER TABLE `form_clinical_notes` ADD COLUMN `clinical_notes_category` varchar(100) DEFAULT NULL AFTER `clinical_notes_type`;
```
rec avoid using `AFTER` in clauses just in case the column ever gets removed from schema in future (then it will through an error). btw, ok for the stuff you did above regarding `id` since that column will never go away :)

adunsulag replied:
K, this latest push should fix that.  I hadn't thought about future upgrades which makes sense.  I usually prefer to have my columns in a specific order.  This has been helpful in the past with dynamic queries and things like CSV imports.  However, with our upgrade script process I could see how this breaks everything.  I've made the change.

bradymiller commented on sql/database.sql:
```diff
@@ -8519,11 +8533,12 @@ CREATE TABLE `uuid_mapping` (
   `created` timestamp NULL,
   PRIMARY KEY (`id`),
-  KEY `uuid` (`uuid`),
+  UNIQUE KEY `uuid` (`uuid`),
```
Any advantage to this?
Purposefully made it not unique for use case of one uuid to represent a group of target uuid (that use case does not exist, but figured it would at some point)

adunsulag replied:
My thought is to make it unique so we don't inadvertently create data integrity issues.  If we want one uuid to represent a group of uuids we can remove the unique constraint at that point.

bradymiller replied:
sounds reasonable. then should also add this change to the upgrade script and update the following comment (to tune of if need to map one uuid to a group of uuids, then need to remove the unique contraint):
https://github.com/openemr/openemr/blob/master/src/Common/Uuid/UuidMapping.php#L23-L24

bradymiller replied:
actually, i don't know if we have a way to do this in the upgrade script (we can check if a key exists, but pretty sure can't check if a key is unique etc.).

bradymiller replied:
if we don't have a way to add to upgrade (and then a way to add unique key in future when needed), then may be best to let this sit as is (ie. not unique).

adunsulag replied:
K, I'll remove the unique constraint

adunsulag replied:
K, just pushed this fix.

bradymiller commented on sql/6_0_0-to-6_1_0_upgrade.sql:
```diff
@@ -687,3 +687,87 @@ ALTER TABLE `form_clinical_notes` ADD `note_related_to` TEXT COMMENT 'Reference
+
+#IfNotRow list_options list_id Clinical_Note_Category
+INSERT INTO `list_options`(`list_id`, `option_id`, `title`, `seq`, `is_default`, `option_value`, `mapping`, `notes`, `codes`, `toggle_setting_1`, `toggle_setting_2`, `activity`, `subtype`, `edit_options`, `timestamp`)
+VALUES
```
also need the entry for the list itself:
```
('lists','Clinical_Note_Category','Clinical Note Category',1,0,0,'','',0,0,0,1,'',1,NOW()),
```

adunsulag replied:
Added this in.

bradymiller commented on sql/database.sql:
```diff
@@ -10737,11 +10752,17 @@ INSERT INTO `list_options` (`list_id`, `option_id`, `title`, `seq`, `is_default`
 INSERT INTO `list_options` (`list_id`, `option_id`, `title`, `seq`, `is_default`, `option_value`, `mapping`, `notes`, `codes`, `toggle_setting_1`, `toggle_setting_2`, `activity`, `subtype`, `edit_options`) VALUES ('Clinical_Note_Type','pathology_report_narrative','Pathology Report Narrative',100,0,0,'','','',0,0,1,'',1);
 
+INSERT INTO `list_options`(`list_id`, `option_id`, `title`, `seq`, `is_default`, `option_value`, `mapping`, `notes`, `codes`, `toggle_setting_1`, `toggle_setting_2`, `activity`, `subtype`, `edit_options`, `timestamp`)
+VALUES
```
also need the entry for the list itself:
```
('lists','Clinical_Note_Category','Clinical Note Category',1,0,0,'','',0,0,0,1,'',1,NOW()),
```

adunsulag replied:
Thanks, I added this in.

bradymiller commented on sql/database.sql:
```diff
@@ -2188,8 +2196,12 @@ CREATE TABLE `form_vitals` (
   `external_id` VARCHAR(20) DEFAULT NULL,
+  `ped_weight_height` float(4,1) default '0.0',
+  `ped_bmi` float(4,1) default '0.0',
+  `ped_head_circ` float(4,1) default '0.0',
```
not sure if this is a meaningful difference, but in the upgrade sql and table sql, default is 0.00, but here is 0.0

adunsulag replied:
I changed the values to be 0.00.

bradymiller commented:
hi @adunsulag , completed another review and only noted several database issues (also looks like some merge conflicts have cropped up). Should be ready for codebase after addressed. thanks!

adunsulag commented:
@bradymiller Everything should be fixed with the tests now passing.  Hope you haven't died of heat stroke!  Thanks for reviewing everything.  Let me know if there is anything else and I can get those resolved.

adunsulag commented:
@bradymiller K brady, I fixed that unique constraint on uuid_mapping.  I went through all of your review comments again and I believe I have everything complete.  I'm ready for you to bring this into master whenever you are ready.  If I've missed something let me know.

bradymiller commented:
hi @adunsulag , looks great and bringing in (gonna fix a tiny discrepancy in interface/forms/clinical_notes/table.sql after bring in this PR; i don't even know why we have those redundant tables on the forms that are installed by default). thanks for the awesome code! -brady

bradymiller commented:
![2DV](https://user-images.githubusercontent.com/278968/124085325-c54fd580-da04-11eb-8f51-d66a9a7e6355.gif)

adunsulag commented:
@bradymiller Thank you.  Glad to finally get this in haha.

sjpadgett commented:
You know @adunsulag, you are starting to make us mere mortals look bad in front of all the other engineers on the play ground.

Truly, love to see you work. Thanks.

stephenwaite commented:
speak for your self :)

https://user-images.githubusercontent.com/5618372/124125361-868f3f00-da47-11eb-947d-d0e08cb32389.mp4
